File: LeetCode 热题 HOT 100/二叉树/515. 在每个树行中找最大值.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var largestValues = function (root) {
  res = []
  if (root === null) {
    return res
  }

  function getVal(node) {
    if (node.left) {
      points.push(node.left)
    }
    if (node.right) {
      points.push(node.right)
    }
  }
  points = []
  points.push(root)
  while (points.length !== 0) {
    result = []
    size = points.length
    for (let index = 0; index < size; index++) {
      temp = points.shift()
      result.push(temp.val)
      getVal(temp)
    }

    res.push(Math.max.apply(null, result))
  }
  return res
};

File: LeetCode 热题 HOT 100/二叉树/515. 在每个树行中找最大值.py
# 给定一棵二叉树的根节点 root ，请找出该二叉树中每一层的最大值。

# 示例1：
# 输入: root = [1, 3, 2, 5, 3, null, 9]
# 输出: [1, 3, 9]

# 示例2：
# 输入: root = [1, 2, 3]
# 输出: [1, 3]


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from collections import deque


class Solution:
    def largestValues(root):
        ans = []
        if not root:
            return ans

        def getVal(root):
            if root.left:
                points.append(root.left)
            if root.right:
                points.append(root.right)

        points = deque([root])
        while points:
            size = len(points)
            result = []
            for _ in range(size):
                temp = points.popleft()
                result.append(temp.val)
                getVal(temp)
            ans.append(max(result))
        return ans
